### PR TITLE
`gpnf-auto-attach-child-files`: Refactored for greater configuration flexibility.

### PR DESCRIPTION
**Pull request by veryspry**
## Context

⛑️ Ticket(s): https://secure.helpscout.net/conversation/2011825626/38759/

## Summary

- Refactors `gpnf-auto-attach-child-files` into a class.
- Moves configuration logic to constructor args so that users can easily create multiple, unique configurations.
- Adds `child_form_upload_field_ids` config option to limit which child form upload fields are included in a notification.

### CODE REVIEW

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (claygriffiths) [inline gp-nested-forms/gpnf-auto-attach-child-files.php:129]: Just for the sake of consistency with our other snippets, can you put the `array` on the same line as the class name?

e.g.

```suggestion
# Attach all uploads from child forms to *only* the notification with id "63331f3fcf7f0".
// new GPNF_Auto_Attach_Child_Files( array(
//    'notification_ids' => array( '63331f3fcf7f0' ),
// ) );
```

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (claygriffiths) [inline gp-nested-forms/gpnf-auto-attach-child-files.php:32]: FWIW, `array_key_exists( 'notification_ids', $config )` can be safely (no PHP notices) shorted to one of the following:

```php
if ( isset( $config['notification_ids'] ) ) {
```

or

```php
if ( ! empty( $config['notification_ids'] ) ) {
```

or even shorter (no if statement)...

```php
$this->notification_ids = rgar( $config, 'notification_ids', array() );
```

**Review by claygriffiths — COMMENTED** [review]
Awesome work!

I'm going to give this a functional test here shortly. In the meantime, here are a couple of minor suggestions for the code.

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (veryspry) [reply gp-nested-forms/gpnf-auto-attach-child-files.php:32]: Noice!

Unfortunately `rgar` isn't always present since this seems to be instantiated before Gravity Forms but I think one of `isset` or `empty` will work :+1:

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (veryspry) [reply gp-nested-forms/gpnf-auto-attach-child-files.php]: Yup! Done

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (veryspry) [reply gp-nested-forms/gpnf-auto-attach-child-files.php]: Done

Comment (claygriffiths) [reply gp-nested-forms/gpnf-auto-attach-child-files.php]: Good call there! If you ever do need GF functions (or functions from perks), you can hook a method to the `plugins_loaded` action.

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (claygriffiths) [inline gp-nested-forms/gpnf-auto-attach-child-files.php:93]: Typo in this method name 😅

**Review by claygriffiths — APPROVED** [review]
Just noticed a typo in a method name and now PHPCS is barking about a couple of things.

I did a functional test and it looks good! I did notice that it slowed down submission a bit, but I'm not sure if that's an issue with the refactor itself.

I went ahead and approved this pending the two small things mentioned above.

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (veryspry) [reply gp-nested-forms/gpnf-auto-attach-child-files.php:93]: :stuck_out_tongue:

**Comment by veryspry** [comment]
> I did a functional test and it looks good! I did notice that it slowed down submission a bit, but I'm not sure if that's an issue with the refactor itself.

Thanks for doing that! All was looking good during my tests but it's always nice to get an extra set of eyes on it :raised_hands: 

As far as the slowdown goes, my guess is that it isn't due to the refactor as the logic is essentially the same as before. The main difference is the way that configuration options are set

If I had to guess why, database reads are probably the main culprit. For instance:

- getting the form object which queries the form table and the form meta table: https://github.com/gravitywiz/snippet-library/pull/535/files#diff-6dc898ef07f19e6f8f6a1c641d2c979e25e78c5327ea4622eccf6b3eb35fb4c7R65
- getting all entries for each child form: https://github.com/gravitywiz/snippet-library/pull/535/files#diff-6dc898ef07f19e6f8f6a1c641d2c979e25e78c5327ea4622eccf6b3eb35fb4c7R66

We also make a call to `preg_replace()` for each attachment we find which might impact speed noticeably?

That said, PCRE is supposed to cache compiled regex's which should be helpful since we pass the same regex string to each call of `preg_replace()`

https://stackoverflow.com/a/210086

**Comment by veryspry** [comment]
Just waiting on confirmation from a user that this works well for them before merging!

File: gp-nested-forms/gpnf-auto-attach-child-files.php
Comment (veryspry) [reply gp-nested-forms/gpnf-auto-attach-child-files.php]: Fixed

**Comment by veryspry** [comment]
Feature request: 

Support for multiple child forms (in one parent form) whose upload fields have different id's.

For example: Parent Form has two child forms, `A` and `B`. Both `A`'s file upload field has id of `1` and `B`'s has a file upload field with id of `2`.

I'm going to go ahead and add that functionality to this snippet before merging!

**Comment by veryspry** [comment]
Feature request: 

Support for multiple child forms (in one parent form) whose upload fields have different id's.

For example: Parent Form has two child forms, `A` and `B`. Both `A`'s file upload field has id of `1` and `B`'s has a file upload field with id of `2`.

I'm going to go ahead and add that functionality to this snippet before merging!

**Comment by veryspry** [comment]
@claygriffiths just a heads up that I changed the API here a bit so that this works with more complex child form configurations.

See here for more details: https://github.com/gravitywiz/snippet-library/pull/535#issuecomment-1265688832

**Review by claygriffiths — APPROVED** [review]

**Comment by veryspry** [comment]
We're just waiting for a user to confirm this works

**Comment by veryspry** [comment]
This is confirmed to work as intended. I'm mergin' it!

**Pull request merged**
